### PR TITLE
Concurrency fixes, DOM delegates exception handler + emit event

### PR DESCRIPTION
**Pull request by georgiosd**
Alright, I was able to reproduce this after all.

It seems to be the result of multiple JS executions happening, I guess at the same time (so it's way after creating the context).

Stack trace:

```
 	mscorlib.dll!System.ThrowHelper.ThrowArgumentException(System.ExceptionResource resource)	Unknown
 	mscorlib.dll!System.Collections.Generic.Dictionary<System.Type, Jint.Native.Object.ObjectInstance>.Insert(System.Type key, Jint.Native.Object.ObjectInstance value, bool add)	Unknown
>	AngleSharp.Scripting.JavaScript.dll!AngleSharp.Scripting.JavaScript.PrototypeCache.GetOrCreate(System.Type type = {Name = "HtmlDivElement" FullName = "AngleSharp.Html.Dom.HtmlDivElement"}, System.Func<System.Type, Jint.Native.Object.ObjectInstance> creator = {Method = {System.Reflection.RuntimeMethodInfo}}) Line 27	C#
 	AngleSharp.Scripting.JavaScript.dll!AngleSharp.Scripting.JavaScript.EngineInstance.GetDomPrototype(System.Type type = {Name = "HtmlDivElement" FullName = "AngleSharp.Html.Dom.HtmlDivElement"}) Line 84	C#
 	AngleSharp.Scripting.JavaScript.dll!AngleSharp.Scripting.JavaScript.DomNodeInstance.DomNodeInstance(AngleSharp.Scripting.JavaScript.EngineInstance engine = {AngleSharp.Scripting.JavaScript.EngineInstance}, object value = {AngleSharp.Html.Dom.HtmlDivElement}) Line 19	C#
 	AngleSharp.Scripting.JavaScript.dll!AngleSharp.Scripting.JavaScript.EngineInstance.CreateInstance(object obj = {AngleSharp.Html.Dom.HtmlDivElement}) Line 101	C#
 	AngleSharp.Scripting.JavaScript.dll!AngleSharp.Scripting.JavaScript.ReferenceCache.GetOrCreate(object obj = {AngleSharp.Html.Dom.HtmlDivElement}, System.Func<object, AngleSharp.Scripting.JavaScript.DomNodeInstance> creator = {Method = {System.Reflection.RuntimeMethodInfo}}) Line 21	C#
 	AngleSharp.Scripting.JavaScript.dll!AngleSharp.Scripting.JavaScript.EngineInstance.GetDomNode(object obj = {AngleSharp.Html.Dom.HtmlDivElement}) Line 79	C#
 	AngleSharp.Scripting.JavaScript.dll!AngleSharp.Scripting.JavaScript.Extensions.ToJsValue(object obj = {AngleSharp.Html.Dom.HtmlDivElement}, AngleSharp.Scripting.JavaScript.EngineInstance engine = {AngleSharp.Scripting.JavaScript.EngineInstance}) Line 52	C#
 	AngleSharp.Scripting.JavaScript.dll!AngleSharp.Scripting.JavaScript.DomFunctionInstance.Call(Jint.Native.JsValue thisObject = {Jint.Native.JsValue}, Jint.Native.JsValue[] arguments = {Jint.Native.JsValue[0]}) Line 34	C#
 	jint.dll!Jint.Runtime.ExpressionInterpreter.EvaluateAssignmentExpression(Jint.Parser.Ast.AssignmentExpression assignmentExpression = {Jint.Parser.Ast.AssignmentExpression})	Unknown
 	jint.dll!Jint.Runtime.ExpressionInterpreter.EvaluateConditionalExpression(Jint.Parser.Ast.ConditionalExpression conditionalExpression)	Unknown
 	jint.dll!Jint.Runtime.StatementInterpreter.ExecuteExpressionStatement(Jint.Parser.Ast.ExpressionStatement expressionStatement)	Unknown
 	jint.dll!Jint.Engine.ExecuteStatement(Jint.Parser.Ast.Statement statement)	Unknown
 	jint.dll!Jint.Runtime.StatementInterpreter.ExecuteStatementList(System.Collections.Generic.IEnumerable<Jint.Parser.Ast.Statement> statementList)	Unknown
 	jint.dll!Jint.Engine.ExecuteStatement(Jint.Parser.Ast.Statement statement)	Unknown
 	jint.dll!Jint.Native.Function.ScriptFunctionInstance.Call(Jint.Native.JsValue thisArg, Jint.Native.JsValue[] arguments)	Unknown
 	jint.dll!Jint.Runtime.ExpressionInterpreter.EvaluateCallExpression(Jint.Parser.Ast.CallExpression callExpression)	Unknown
 	jint.dll!Jint.Runtime.ExpressionInterpreter.EvaluateAssignmentExpression(Jint.Parser.Ast.AssignmentExpression assignmentExpression = {Jint.Parser.Ast.AssignmentExpression})	Unknown
 	jint.dll!Jint.Runtime.StatementInterpreter.ExecuteForStatement(Jint.Parser.Ast.ForStatement forStatement = {Jint.Parser.Ast.ForStatement})	Unknown
 	jint.dll!Jint.Engine.ExecuteStatement(Jint.Parser.Ast.Statement statement)	Unknown
 	jint.dll!Jint.Runtime.StatementInterpreter.ExecuteStatementList(System.Collections.Generic.IEnumerable<Jint.Parser.Ast.Statement> statementList)	Unknown
 	jint.dll!Jint.Engine.ExecuteStatement(Jint.Parser.Ast.Statement statement)	Unknown
 	jint.dll!Jint.Native.Function.ScriptFunctionInstance.Call(Jint.Native.JsValue thisArg, Jint.Native.JsValue[] arguments)	Unknown
 	AngleSharp.Scripting.JavaScript.dll!AngleSharp.Scripting.JavaScript.DomDelegates.ToListener.AnonymousMethod__0(object obj = {AngleSharp.Html.Dom.HtmlHtmlElement}, AngleSharp.Dom.Events.Event ev = {AngleSharp.Dom.Events.Event}) Line 29	C#
 	AngleSharp.dll!AngleSharp.Dom.EventTarget.InvokeEventListener(AngleSharp.Dom.Events.Event ev = {AngleSharp.Dom.Events.Event}) Line 124	C#
 	AngleSharp.dll!AngleSharp.Dom.Events.Event.CallListeners(AngleSharp.Dom.IEventTarget target = {AngleSharp.Html.Dom.HtmlHtmlElement}) Line 250	C#
 	AngleSharp.dll!AngleSharp.Dom.Events.Event.DispatchAt(System.Collections.Generic.IEnumerable<AngleSharp.Dom.IEventTarget> targets = {System.Linq.Enumerable.<ReverseIterator>d__74<AngleSharp.Dom.IEventTarget>}) Line 257	C#
 	AngleSharp.dll!AngleSharp.Dom.Events.Event.Dispatch(AngleSharp.Dom.IEventTarget target = {AngleSharp.Html.Dom.HtmlScriptElement}) Line 227	C#
 	AngleSharp.dll!AngleSharp.Dom.EventTargetExtensions.FireSimpleEvent(AngleSharp.Dom.IEventTarget target = {AngleSharp.Html.Dom.HtmlScriptElement}, string eventName = "load", bool bubble = false, bool cancelable = false) Line 28	C#
 	AngleSharp.dll!AngleSharp.Io.Processors.ScriptRequestProcessor.FireLoadEvent() Line 156	C#
 	AngleSharp.dll!AngleSharp.Browser.EventLoopExtensions.Enqueue.AnonymousMethod__0(System.Threading.CancellationToken c = IsCancellationRequested = false) Line 20	C#
 	AngleSharp.dll!AngleSharp.Browser.TaskEventLoop.TaskEventLoopEntry.Run.AnonymousMethod__0() Line 144	C#
 	mscorlib.dll!System.Threading.Tasks.Task.Execute()	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecuteWithThreadLocal(ref System.Threading.Tasks.Task currentTaskSlot = Id = 1077, Status = Running, Method = "Void <Run>b__0()")	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecuteEntry(bool bPreventDoubleExecution)	Unknown
 	mscorlib.dll!System.Threading.ThreadPoolWorkQueue.Dispatch()	Unknown
```

The same happens with `ReferenceCache`. Stack trace:

 	mscorlib.dll!System.ThrowHelper.ThrowArgumentException(System.ExceptionResource resource)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.ConditionalWeakTable<System.__Canon, System.__Canon>.Add(System.__Canon key, System.__Canon value)	Unknown
	AngleSharp.Scripting.JavaScript.dll!AngleSharp.Scripting.JavaScript.ReferenceCache.GetOrCreate(object obj = {AngleSharp.Html.Dom.HtmlBodyElement}, System.Func<object, AngleSharp.Scripting.JavaScript.DomNodeInstance> creator = {Method = {System.Reflection.RuntimeMethodInfo}}) Line 22	C#
 	AngleSharp.Scripting.JavaScript.dll!AngleSharp.Scripting.JavaScript.EngineInstance.GetDomNode(object obj = {AngleSharp.Html.Dom.HtmlBodyElement}) Line 79	C#
 	AngleSharp.Scripting.JavaScript.dll!AngleSharp.Scripting.JavaScript.Extensions.ToJsValue(object obj = {AngleSharp.Html.Dom.HtmlBodyElement}, AngleSharp.Scripting.JavaScript.EngineInstance engine = {AngleSharp.Scripting.JavaScript.EngineInstance}) Line 52	C#
 	AngleSharp.Scripting.JavaScript.dll!AngleSharp.Scripting.JavaScript.DomFunctionInstance.Call(Jint.Native.JsValue thisObject = {Jint.Native.JsValue}, Jint.Native.JsValue[] arguments = {Jint.Native.JsValue[0]}) Line 34	C#
 	jint.dll!Jint.Runtime.ExpressionInterpreter.EvaluateAssignmentExpression(Jint.Parser.Ast.AssignmentExpression assignmentExpression = {Jint.Parser.Ast.AssignmentExpression})	Unknown
 	jint.dll!Jint.Runtime.ExpressionInterpreter.EvaluateConditionalExpression(Jint.Parser.Ast.ConditionalExpression conditionalExpression)	Unknown
 	jint.dll!Jint.Runtime.StatementInterpreter.ExecuteExpressionStatement(Jint.Parser.Ast.ExpressionStatement expressionStatement)	Unknown
 	jint.dll!Jint.Engine.ExecuteStatement(Jint.Parser.Ast.Statement statement)	Unknown
 	jint.dll!Jint.Runtime.StatementInterpreter.ExecuteStatementList(System.Collections.Generic.IEnumerable<Jint.Parser.Ast.Statement> statementList)	Unknown
 	jint.dll!Jint.Engine.ExecuteStatement(Jint.Parser.Ast.Statement statement)	Unknown
 	jint.dll!Jint.Native.Function.ScriptFunctionInstance.Call(Jint.Native.JsValue thisArg, Jint.Native.JsValue[] arguments)	Unknown
 	jint.dll!Jint.Runtime.ExpressionInterpreter.EvaluateCallExpression(Jint.Parser.Ast.CallExpression callExpression)	Unknown
 	jint.dll!Jint.Runtime.ExpressionInterpreter.EvaluateAssignmentExpression(Jint.Parser.Ast.AssignmentExpression assignmentExpression = {Jint.Parser.Ast.AssignmentExpression})	Unknown
 	jint.dll!Jint.Runtime.StatementInterpreter.ExecuteForStatement(Jint.Parser.Ast.ForStatement forStatement = {Jint.Parser.Ast.ForStatement})	Unknown
 	jint.dll!Jint.Engine.ExecuteStatement(Jint.Parser.Ast.Statement statement)	Unknown
 	jint.dll!Jint.Runtime.StatementInterpreter.ExecuteStatementList(System.Collections.Generic.IEnumerable<Jint.Parser.Ast.Statement> statementList)	Unknown
 	jint.dll!Jint.Engine.ExecuteStatement(Jint.Parser.Ast.Statement statement)	Unknown
 	jint.dll!Jint.Native.Function.ScriptFunctionInstance.Call(Jint.Native.JsValue thisArg, Jint.Native.JsValue[] arguments)	Unknown
 	AngleSharp.Scripting.JavaScript.dll!AngleSharp.Scripting.JavaScript.DomDelegates.ToListener.AnonymousMethod__0(object obj = {AngleSharp.Html.Dom.HtmlHtmlElement}, AngleSharp.Dom.Events.Event ev = {AngleSharp.Dom.Events.Event}) Line 29	C#
 	AngleSharp.dll!AngleSharp.Dom.EventTarget.InvokeEventListener(AngleSharp.Dom.Events.Event ev = {AngleSharp.Dom.Events.Event}) Line 124	C#
 	AngleSharp.dll!AngleSharp.Dom.Events.Event.CallListeners(AngleSharp.Dom.IEventTarget target = {AngleSharp.Html.Dom.HtmlHtmlElement}) Line 250	C#
 	AngleSharp.dll!AngleSharp.Dom.Events.Event.DispatchAt(System.Collections.Generic.IEnumerable<AngleSharp.Dom.IEventTarget> targets = {System.Linq.Enumerable.<ReverseIterator>d__74<AngleSharp.Dom.IEventTarget>}) Line 257	C#
 	AngleSharp.dll!AngleSharp.Dom.Events.Event.Dispatch(AngleSharp.Dom.IEventTarget target = {AngleSharp.Html.Dom.HtmlScriptElement}) Line 227	C#
 	AngleSharp.dll!AngleSharp.Dom.EventTargetExtensions.FireSimpleEvent(AngleSharp.Dom.IEventTarget target = {AngleSharp.Html.Dom.HtmlScriptElement}, string eventName = "load", bool bubble = false, bool cancelable = false) Line 28	C#
 	AngleSharp.dll!AngleSharp.Io.Processors.ScriptRequestProcessor.FireLoadEvent() Line 156	C#
 	AngleSharp.dll!AngleSharp.Browser.EventLoopExtensions.Enqueue.AnonymousMethod__0(System.Threading.CancellationToken c = IsCancellationRequested = false) Line 20	C#
 	AngleSharp.dll!AngleSharp.Browser.TaskEventLoop.TaskEventLoopEntry.Run.AnonymousMethod__0() Line 144	C#
 	mscorlib.dll!System.Threading.Tasks.Task.Execute()	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecuteWithThreadLocal(ref System.Threading.Tasks.Task currentTaskSlot = Id = 1097, Status = Running, Method = "Void <Run>b__0()")	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.ExecuteEntry(bool bPreventDoubleExecution)	Unknown
 	mscorlib.dll!System.Threading.ThreadPoolWorkQueue.Dispatch()	Unknown

```

All this on google.com!

### CODE REVIEW

**Comment by georgiosd** [comment]
I think I did the error event as you wanted it. Cheers!

File: src/AngleSharp.Scripting.JavaScript/AngleSharp.Scripting.JavaScript.csproj [hunk 1]
```diff
@@ -42,6 +42,9 @@
       <Private>True</Private>
     </Reference>
     <Reference Include="System" />
+    <Reference Include="System.Collections.Concurrent">
+      <HintPath>..\..\..\..\..\..\..\..\..\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETPortable\v4.5\Profile\Profile111\System.Collections.Concurrent.dll</HintPath>
```
Comment (FlorianRappl) [inline src/AngleSharp.Scripting.JavaScript/AngleSharp.Scripting.JavaScript.csproj]: Looks strange.

**Comment by FlorianRappl** [comment]
Looks wonderful! 🍺

File: src/AngleSharp.Scripting.JavaScript/AngleSharp.Scripting.JavaScript.csproj (hunk 1 above)
Comment (georgiosd) [inline src/AngleSharp.Scripting.JavaScript/AngleSharp.Scripting.JavaScript.csproj]: Yeah I'm not sure why it did like that. I can try removing the hint and see if it builds OK.

File: src/AngleSharp.Scripting.JavaScript/AngleSharp.Scripting.JavaScript.csproj [hunk 2]
```diff
@@ -87,5 +82,11 @@
   <ItemGroup>
     <None Include="packages.config" />
   </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\AngleSharp\src\AngleSharp\AngleSharp.Core.csproj">
```
Comment (FlorianRappl) [inline src/AngleSharp.Scripting.JavaScript/AngleSharp.Scripting.JavaScript.csproj]: I think now AngleSharp dropped from the list of NuGet packages.

Comment (georgiosd) [inline src/AngleSharp.Scripting.JavaScript/AngleSharp.Scripting.JavaScript.csproj]: Woops!

Comment (georgiosd) [inline src/AngleSharp.Scripting.JavaScript/AngleSharp.Scripting.JavaScript.csproj]: This is a problem for me - how do you develop code for this branch?? I mean 0.10.0.0 doesn't exist in the nuget repo so the build fails...

**Comment by georgiosd** [comment]
Ok I cleaned up the HintPath mishap.

**Comment by FlorianRappl** [comment]
Looks great!

**Comment by briman0094** [comment]
This commit doesn't build against the `AngleSharp:devel` branch at the time of writing. It seems to call an overload of `Event.Init` on `DomDelegates.cs:45` that doesn't exist. The current state of `AngleSharp.Scripting:devel` also references `AngleSharp.0.10.0` on NuGet which doesn't exist currently, and I'm guessing that may be related to the disjunction between the `Event.Init` call.

**Comment by georgiosd** [comment]
@briman0094 the 0.10.0 business was there before I came along so you'll have to ask @FlorianRappl  about that. I'm not sure how he develops with it, I had to directly reference the `AngleSharp.Core` csproj for it.

Re the `ErrorEvent.Init` I'm not sure why it's missing in this tree, but shouldn't, will look into it now.

**Comment by briman0094** [comment]
That's what I'm currently linking against...I have the lastest version of the `devel` branch checked out and I'm referencing that project in `AngleSharp.Scripting`. I can check again once I get back to work on Monday to make sure I definitely have the latest, but I'm nearly positive I do.

**Comment by georgiosd** [comment]
@briman0094 it's because my main project PR isn't merged yet. See https://github.com/AngleSharp/AngleSharp/pull/562

I'm waiting on some feedback from @FlorianRappl before I can send the final commits. Meantime you can checkout my branch i guess.

**Pull request merged**